### ISSUE TITLE
Cannot move `access` from `google_bigquery_dataset`

### ISSUE THREAD

**Issue by olib963** [hashicorp/terraform-provider-google#8165]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Currently terraform 0.13.5 and google provider 3.44.0, however is still an issue when attempted with 0.14.4 and 3.51.1 respectively.

### Affected Resource(s)

- google_bigquery_dataset

### Terraform Configuration Files

```tf
# Before
resource "google_bigquery_dataset" "private" {
  dataset_id    = "test-dataset"
  friendly_name = "Test"
  description   = "Testing BQ Permissions"
  location      = "EU"

   access {
    role          = "OWNER"
    user_by_email = google_service_account.bqowner.email
  }

  access {
    role   = "READER"
    domain = "hashicorp.com"
  }
}

resource "google_service_account" "bqowner" {
  account_id = "bqowner"
}
```

```tf
# After
resource "google_bigquery_dataset" "private" {
  dataset_id    = "test-dataset"
  friendly_name = "Test"
  description   = "Testing BQ Permissions"
  location      = "EU"
}

#resource "google_bigquery_dataset_access" "reader" {
#  dataset_id    = google_bigquery_dataset.private.dataset_id
#  role          = "READER"
#  domain = "hashicorp.com"
#}

#resource "google_bigquery_dataset_access" "owner" {
#  dataset_id    = google_bigquery_dataset.private.dataset_id
#  role          = "OWNER"
#  user_by_email = google_service_account.bqowner.email
#}

resource "google_service_account" "bqowner" {
  account_id = "bqowner"
}
```
### Expected Behavior

I should be able to move the access definition from the dataset itself to separate resources. The reason for doing this is that currently we have a lot of `access` blocks and a change in any of them results in an ugly plan that removes _all_ `access` blocks, even the ones that don't change, then adds all the correct blocks. I thought that having each access block live on its own would make the changes in state much clearer.

### Actual Behavior

Removing all `access` blocks from a dataset however results in an empty plan and you cannot set `access = []` since that results in a terraform error asking if "you meant you define an "access" block". This means we are unable to transition to the `google_bigquery_dataset_access` resource.

### Steps to Reproduce

1. `terraform apply` the "before" terraform file
2. `terraform plan` or `terraform apply` the "after" file to see that no changes are detected

Ideally there would then be a step 3: uncomment new resources and `terraform apply` again to get access back.

**Comment by slevenick** [comment]
Hey @olib963 does adding `ignore_changes` work for your use case? It should cause Terraform to ignore any changes to the access block, preventing the diff on the main dataset resource.

https://www.terraform.io/docs/configuration/meta-arguments/lifecycle.html#ignore_changes

I don't believe there is a better way short of destroying and recreating the dataset with access unspecified

**Comment by olib963** [comment]
I can't ignore changes since the `google_bigquery_dataset_access` currently fails to create because the access is already defined on the dataset itself. I'd like to not destroy and recreate since there's a lot of data in there. I'd have to come up with some process to stop writing new data and back it up then copy it back in.

Do we know why it is the case that removing all access blocks does not show a change in the desired state? For example if I change the resource to 

```tf
resource "google_bigquery_dataset" "private" {
  dataset_id    = "test-dataset"
  friendly_name = "Test"
  description   = "Testing BQ Permissions"
  location      = "EU"

   access {
    role          = "OWNER"
    user_by_email = google_service_account.bqowner.email
  }
}
```

It completely understands it needs to remove the `READER` access, but removing both leads to an empty plan.

**Comment by slevenick** [comment]
Yeah, when you remove the `access` block entirely it interprets that as "I don't care about what access is" rather than "access should be empty". This is important because once you manage access with the `google_bigquery_dataset_access` resource, we don't want the bigquery dataset to revert any changes. Is this empty plan a problem? I'd suggest that you could remove the access manually and recreate using the `google_bigquery_dataset_access` resource. It's a bit of an annoying step, but could get you to the world where the dataset does not manage access, and the `access` resource does.


By moving to using the `google_bigquery_dataset_access` resource you are moving to a non-authoritative permissions model, meaning that there could be additional access/permissions on the dataset that will not be removed by Terraform. If you want to continue using an authoritative model you could look into using https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_dataset_iam
Note: this will not allow some advanced usages for datasets

**Comment by slevenick** [comment]
Also, I believe this feature request is for something that would help moving between these resources: https://github.com/hashicorp/terraform-provider-google/issues/7486

**Comment by olib963** [comment]
Ah OK, that makes sense now. I could look into manually removing the access myself in the console, but we don't have write permissions intentionally. I will use that as a backup.

Sadly I can't use the IAM resources as I need authorised views. I'd like to keep the authoritative model if possible, the current setup is just causing a bit of a headache. Perhaps I will comment a slightly different view on the problem.

If I apply the above "before" dataset and then change `hashicorp.com` to `hashicorp.co.uk` this is the plan I get:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # google_bigquery_dataset.private will be updated in-place
  ~ resource "google_bigquery_dataset" "private" {
        id                              = "projects/billing-and-statements-dev/datasets/test_dataset"
        # (13 unchanged attributes hidden)

      - access {
          - role          = "OWNER" -> null
          - user_by_email = "bqowner@billing-and-statements-dev.iam.gserviceaccount.com" -> null
        }
      + access {
          + domain = "hashicorp.co.uk"
          + role   = "READER"
        }
      - access {
          - domain = "hashicorp.com" -> null
          - role   = "READER" -> null
        }
      + access {
          + role          = "OWNER"
          + user_by_email = "bqowner@billing-and-statements-dev.iam.gserviceaccount.com"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

```

What I would like it to be is something like this:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # google_bigquery_dataset.private will be updated in-place
  ~ resource "google_bigquery_dataset" "private" {
        id                              = "projects/billing-and-statements-dev/datasets/test_dataset"
        # (13 unchanged attributes hidden)

      ~ access {
          ~ domain = "hashicorp.com" -> "hashicorp.co.uk"
          role   = "READER
        }
      access {
          role          = "OWNER"
          user_by_email = "bqowner@billing-and-statements-dev.iam.gserviceaccount.com"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Similar to how when I change a value on a container in a kubernetes deployment the plan will only show that change, not deleting all containers then adding new containers.

The problem is the current output is hiding the diffs which makes PRs harder to review and daily checks for when people have changed permissions outside of terraform hard to parse.

Is this possible to do?
